firedrake/preconditioners/fdm.py

```diff
@@ -136,73 +187,960 @@ def interp_nullspace(I, nsp):
+
+        value_size = Vbig.value_size
+        if value_size != 1:
+            fdofs = numpy.add.outer(value_size * fdofs, numpy.arange(value_size, dtype=fdofs.dtype))
+        dofs = numpy.arange(value_size * Vbig.finat_element.space_dimension(), dtype=fdofs.dtype)
+        idofs = numpy.setdiff1d(dofs, fdofs, assume_unique=True)
+        self.ises = tuple(PETSc.IS().createGeneral(indices, comm=PETSc.COMM_SELF) for indices in (idofs, fdofs))
+        self.submats = [None for _ in range(7)]
+
+        self.reference_tensor_on_diag = dict()
+        self.get_static_condensation = dict()
+        if Vfacet and use_static_condensation:
+            # If we are in a facet space, we build the Schur complement on its diagonal block
+            self.reference_tensor_on_diag[Vfacet] = self.assemble_reference_tensor(Vbig)
+            self.get_static_condensation[Vfacet] = lambda A: condense_element_mat(A, self.ises[0], self.ises[1], self.submats)
+
+        elif len(fdofs) and V.finat_element.formdegree == 0:
+            # If we are in H(grad), we just pad with zeros on the statically-condensed pattern
+            i1 = PETSc.IS().createGeneral(dofs, comm=PETSc.COMM_SELF)
+            self.get_static_condensation[V] = lambda Ae: condense_element_pattern(Ae, self.ises[0], i1, self.submats)
+
+        @PETSc.Log.EventDecorator("FDMGetIndices")
+        def cell_to_global(lgmap, cell_to_local, cell_index, result=None):
+            # Be careful not to create new arrays
+            result = cell_to_local(cell_index, result=result)
+            return lgmap.apply(result, result=result)
+
+        # Create data structures needed for assembly
+        self.cell_to_global = dict()
+        self.lgmaps = dict()
+        bc_rows = dict()
+        for Vsub in V:
+            lgmap = Vsub.local_to_global_map([bc.reconstruct(V=Vsub, g=0) for bc in bcs])
+            bsize = Vsub.dof_dset.layout_vec.getBlockSize()
+            cell_to_local, nel = extrude_node_map(Vsub.cell_node_map(), bsize=bsize)
+            self.cell_to_global[Vsub] = partial(cell_to_global, lgmap, cell_to_local)
+            self.lgmaps[Vsub] = lgmap
+
+            own = Vsub.dof_dset.layout_vec.getLocalSize()
+            bdofs = numpy.nonzero(lgmap.indices[:own] < 0)[0].astype(PETSc.IntType)
+            bc_rows[Vsub] = Vsub.dof_dset.lgmap.apply(bdofs, result=bdofs)
+
+        coefficients, assembly_callables = self.assemble_coef(J, form_compiler_parameters)
+        coeffs = [coefficients.get(k) for k in ("beta", "alpha")]
+        cmaps = [extrude_node_map(ck.cell_node_map())[0] for ck in coeffs]
+
+        @PETSc.Log.EventDecorator("FDMGetCoeffs")
+        def get_coeffs(e, result=None):
+            # Get vector for betas and alphas on a cell
+            vals = []
+            for k, (coeff, cmap) in enumerate(zip(coeffs, cmaps)):
+                get_coeffs.indices[k] = cmap(e, result=get_coeffs.indices[k])
```

Comment:
This looks like you are attaching attributes to a function pointer from within the function itself. Can you explain what is going on? I think there may be a clearer way to achieve what you want.

Reply:
replaced this with variables external to the function